### PR TITLE
feat(reachableservices): implement reachable services for Mesh*Service and auto-rechable services for MeshExternalService

### PR DESCRIPTION
**Pull request by lukidzi**
### Checklist prior to review

closes https://github.com/kumahq/kuma/issues/10602
Impl https://github.com/kumahq/kuma/pull/10737
Impl https://github.com/kumahq/kuma/issues/10508
Solves https://github.com/kumahq/kuma/issues/5041 for reachableBackends

This PR provides a way to define reachable services for MeshService and MeshExternalService. Also, it introduces a way to support auto reachable services for a MeshService.

* added new annotation `kuma.io/reachable-backend-refs` and dataplane field `transparentProxying.rechableBackends.refs` which allows providing definition of Mesh*Service reachable for a dataplane in a structured way
* In MeshContext I am building a map for labels to Mesh(service, so once the user defines that wants reach all services in a namespace `k8s.kuma.io/namespace: demo` can find all MeshServices that have this label (should we consider limiting allowed tags?)
* we add outbound based on backendRef and non backend ref, in the case of a cluster with MeshService and another without MeshService we add them based of existence of annotations
  * if reachableSevices present we don't add backend refs
  * if  reachableBackends present we don't add kuma.io/services
* One case that not fully sure how to handle expressed by the test https://github.com/kumahq/kuma/pull/10869/files#diff-e5fd492c9fca43317e405258365acd5ac7b6c69f43186739de3919c00329e1deR333

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/xds/sync/dataplane_proxy_builder.go [hunk 1]
```diff
@@ -123,6 +126,10 @@ func (p *DataplaneProxyBuilder) resolveVIPOutbounds(meshContext xds_context.Mesh
 	var outbounds []*mesh_proto.Dataplane_Networking_Outbound
 	for _, outbound := range meshContext.VIPOutbounds {
 		if outbound.BackendRef == nil { // reachable services does not work with backend ref yet.
+			// TODO(lukidzi): should we support both Meshservice and kuma.io simultaneously?
```
Comment (jakubdyszkiewicz) [inline pkg/xds/sync/dataplane_proxy_builder.go:129]: yes.

File: pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+					Items: trimNotSupportedTags(mtps, dpTags),
+				},
+			},
+		}
+		// build artificial dpp for matching
+		dp := mesh.NewDataplaneResource()
+
+		dp.Spec = &mesh_proto.Dataplane{
+			Networking: &mesh_proto.Dataplane_Networking{
+				Address: "1.1.1.1",
```
Comment (jijiechen) [inline pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go:71]: The values `1.1.1.1` and `1234` here look like placeholders. Why don't they impact the actual feature?

File: pkg/plugins/policies/meshtrafficpermission/graph/reachable_graph.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+func NewGraph(rules map[string]core_rules.Rules, backendRules map[backends.BackendKey]core_rules.Rules) *Graph {
+	return &Graph{
+		rules:        rules,
+		backendRules: backendRules,
+	}
+}
+
+func (r *Graph) CanReach(fromTags map[string]string, toTags map[string]string) bool {
+	if _, crossMeshTagExist := toTags[mesh_proto.MeshTag]; crossMeshTagExist {
```
Comment (jijiechen) [inline pkg/plugins/policies/meshtrafficpermission/graph/reachable_graph.go:33]: Just a question, can `fromTags` also have key `mesh_proto.MeshTag` and it equals to that in `toTags`?

**Comment by jijiechen** [comment]
It looks good to me. Thanks.

**Review by jijiechen — APPROVED** [review]

File: pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go (hunk 2 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go:71]: It's only a placeholder no relevant for actual matching

File: pkg/plugins/policies/meshtrafficpermission/graph/reachable_graph.go (hunk 3 above)
Comment (lukidzi) [reply pkg/plugins/policies/meshtrafficpermission/graph/reachable_graph.go:33]: `mesh_proto.MeshTag` is used only for cross mesh traffic and we don't really match it(cross mesh is only on the gateway)

**Comment by Icarus9913** [comment]
Will there be a conflict once the annotation `kuma.io/reachable-backend-refs` and `kuma.io/transparent-proxying-reachable-services` are both used? 
Or we have a priority with the 2 annotations or we can disallow this behavior in webhook?

**Comment by lukidzi** [comment]
> Will there be a conflict once the annotation `kuma.io/reachable-backend-refs` and `kuma.io/transparent-proxying-reachable-services` are both used? Or we have a priority with the 2 annotations or we can disallow this behavior in webhook?

I think we can do following:
1. When only reachable-services or reachable-backends defined then use only one of them and do not include other
2. When both defined use both

File: pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		Name: backendRef.Name,
+	}].Compute(core_rules.SubsetFromTags(fromTags))
+	if rule == nil {
+		return false
+	}
+	action := rule.Conf.(mtp_api.Conf).Action
+	return action == mtp_api.Allow || action == mtp_api.AllowWithShadowDeny
+}
+
+func BuildRules(meshServices []*ms_api.MeshServiceResource, mtps []*mtp_api.MeshTrafficPermissionResource) map[BackendKey]core_rules.Rules {
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go:49]: The duplication here is quite significant. This func could basically just be
```
rules := map[BackendKey]core_rules.Rules{}
for _, ms := range meshServices {
		dpTags := maps.Clone(ms.Spec.Selector.DataplaneTags)
		if origin, ok := core_model.ResourceOrigin(ms.GetMeta()); ok {
			dpTags[mesh_proto.ResourceOriginLabel] = string(origin)
		}
		if ms.GetMeta().GetLabels() != nil && ms.GetMeta().GetLabels()[mesh_proto.ZoneTag] != "" {
			dpTags[mesh_proto.ZoneTag] = ms.GetMeta().GetLabels()[mesh_proto.ZoneTag]
		}

rl, ok := ComputeMtpRulesForTags(dpTags, TrimNotSupportedTags(mtps))
		if !ok {
			continue
		}

		rules[BackendKey{
			Kind: "MeshService",
			Name: ms.Meta.GetName(),
		}] = rl

```
you can then call `ComputeMtpRulesForTags` from both graphs.

File: pkg/plugins/policies/meshtrafficpermission/graph/reachable_backend_refs_graph_test.go [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+		},
+		Entry("MeshSubset by kube namespace", builders.TargetRefMeshSubset(mesh_proto.KubeNamespaceTag, "kuma-demo")),
+		Entry("MeshSubset by kube service name", builders.TargetRefMeshSubset(mesh_proto.KubeServiceTag, "a")),
+		Entry("MeshSubset by kube service port", builders.TargetRefMeshSubset(mesh_proto.KubePortTag, "1234")),
+		Entry("MeshSubset by app and kube namespace", builders.TargetRefMeshSubset("app", "a", mesh_proto.KubeNamespaceTag, "kuma-demo")),
+		Entry("MeshSubset by app and kube service name", builders.TargetRefMeshSubset("app", "a", mesh_proto.KubeServiceTag, "a")),
+		Entry("MeshSubset by app and kube service port", builders.TargetRefMeshSubset("app", "a", mesh_proto.KubePortTag, "1234")),
+	)
+
+	XIt("should match only dp with specific labels", func() {
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/meshtrafficpermission/graph/reachable_backend_refs_graph_test.go:346]: why XIt?

File: pkg/plugins/runtime/k8s/controllers/pod_converter.go [hunk 6]
```diff
@@ -145,6 +147,34 @@ func (p *PodConverter) dataplaneFor(
...
+			for _, ref := range refs.Refs {
+				backendRef := &mesh_proto.Dataplane_Networking_TransparentProxying_ReachableBackendRef{
+					Kind:   ref.Kind,
+					Labels: ref.Labels,
+				}
+				if ref.Port != nil {
+					backendRef.Port = util_proto.UInt32(pointer.Deref(ref.Port))
+				}
+				if ref.Name != nil {
+					backendRef.Name = pointer.Deref(ref.Name)
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/runtime/k8s/controllers/pod_converter.go:166]: pointer.Deref does not make sense when you already checked if pointer is not nil. Did you mean
```
backendRef.Name = pointer.Deref(ref.Name)
backendRef.Namespace = pointer.Deref(ref.Namespace)
```
? without ifs
or
```
backendRef.Name = *ref.Name
```
with ifs

File: pkg/plugins/runtime/k8s/controllers/pod_converter.go [hunk 7]
```diff
@@ -145,6 +147,34 @@ func (p *PodConverter) dataplaneFor(
...
+				}
+				if ref.Namespace != nil {
+					backendRef.Namespace = pointer.Deref(ref.Namespace)
+				}
+				backendRefs = append(backendRefs, backendRef)
+			}
+			dataplane.Networking.TransparentProxying.ReachableBackends = &mesh_proto.Dataplane_Networking_TransparentProxying_ReachableBackends{}
+			if len(backendRefs) > 0 {
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/runtime/k8s/controllers/pod_converter.go:174]: do we need to check that it's greater than 0? can't we just set it?

File: pkg/plugins/runtime/k8s/controllers/testdata/28.pod.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  namespace: demo
+  name: example
+  labels:
+    app: example
+    version: "0.1"
+  annotations:
+    kuma.io/transparent-proxying: "enabled"
+    kuma.io/transparent-proxying-inbound-port: 15006
+    kuma.io/transparent-proxying-outbound-port: 15001
+    kuma.io/reachable-backend-refs: |
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/runtime/k8s/controllers/testdata/28.pod.yaml:11]: already discussed offline, but `kuma.io/reachable-backends`

File: pkg/xds/context/context.go [hunk 9]
```diff
@@ -88,6 +92,74 @@ type ServiceInformation struct {
...
+	}
+	reachableBackends := ReachableBackends{}
+	for _, reachableBackend := range dataplane.Spec.Networking.TransparentProxying.GetReachableBackends().GetRefs() {
+		key := BackendKey{Kind: reachableBackend.Kind}
+		name := ""
+		if reachableBackend.Name != "" {
+			name = reachableBackend.Name
+		}
+		if reachableBackend.Namespace != "" {
+			name += fmt.Sprintf(".%s", reachableBackend.Namespace)
```
Comment (jakubdyszkiewicz) [inline pkg/xds/context/context.go:115]: use `K8sNamespacedNameToCoreName`

File: pkg/xds/context/mesh_context_builder.go [hunk 10]
```diff
@@ -161,18 +161,25 @@ func (m *meshContextBuilder) BuildIfChanged(ctx context.Context, meshName string
 	}
 	meshServices := resources.MeshServices().Items
 	meshServicesByName := make(map[string]*v1alpha1.MeshServiceResource, len(meshServices))
+	meshServicesByLabels := map[string]map[string][]string{}
```
Comment (jakubdyszkiewicz) [inline pkg/xds/context/mesh_context_builder.go:164]: It's not really `meshServicesByLabels` it's mor of `meshServiceNamesByLabelByValue`

I wonder. Do we need `map[string]map[string][]string{}` or could this be
```
meshServiceNamesByLabelValue := map[LabelValue]string
type LabelValue struct {
  Label string
  Value string
}

or maybe
`meshServicesByLabelValue := map[LabelValue]MeshServiceResource`
```

File: pkg/xds/sync/dataplane_proxy_builder.go [hunk 11]
```diff
@@ -136,6 +141,28 @@ func (p *DataplaneProxyBuilder) resolveVIPOutbounds(meshContext xds_context.Mesh
 					continue
 				}
 			}
+		} else {
+			if len(reachableServices) != 0 && reachableBackends == nil {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/sync/dataplane_proxy_builder.go:145]: add explanation in the comment. That we don't want to ruin performance of existing deployments that has reachable services defined already

File: pkg/xds/sync/dataplane_proxy_builder.go [hunk 12]
```diff
@@ -136,6 +141,28 @@ func (p *DataplaneProxyBuilder) resolveVIPOutbounds(meshContext xds_context.Mesh
...
+					Name: outbound.BackendRef.Name,
+					Port: outbound.BackendRef.Port,
+				}
+				// check if there is an entry with specific port or without port
+				if !pointer.Deref(reachableBackends)[backendKey] && !pointer.Deref(reachableBackends)[xds_context.BackendKey{Kind: outbound.BackendRef.Kind, Name: outbound.BackendRef.Name}] {
+					// ignore VIP outbound if reachableServices is defined and not specified
+					// Reachable services takes precedence over reachable services graph.
+					continue
+				}
+			} else if outbound.BackendRef.Kind != "MeshExternalService" {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/sync/dataplane_proxy_builder.go:160]: Yes, but why do we skip MeshExternalService, add a reason in the comment which is that we don't have MTP yet for MEshExternalServices if I understand the code correctly

File: test/e2e/reachableservices/auto_reachable_mesh_services_k8s.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  template: "{{ .DisplayName }}.mesh"
+`, meshName, Config.KumaNamespace, namespace)
+
+	BeforeAll(func() {
+		k8sCluster = NewK8sCluster(NewTestingT(), Kuma1, Silent)
+
+		err := NewClusterSetup().
+			Install(Kuma(config_core.Zone,
+				WithEnv("KUMA_EXPERIMENTAL_AUTO_REACHABLE_SERVICES", "true"),
+				WithEnv("KUMA_EXPERIMENTAL_GENERATE_MESH_SERVICES", "true"),
```
Comment (jakubdyszkiewicz) [inline test/e2e/reachableservices/auto_reachable_mesh_services_k8s.go:43]: I'd also consider `KUMA_EXPERIMENTAL_SKIP_PERSISTED_VIPS`

File: test/e2e/reachableservices/auto_reachable_mesh_services_k8s.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+		Consistently(func(g Gomega) {
+			failures, err := client.CollectFailure(
+				k8sCluster,
+				"second-test-server",
+				"first-test-server.mesh",
+				client.FromKubernetesPod(namespace, "second-test-server"),
+			)
+			g.Expect(err).To(Not(HaveOccurred()))
+			g.Expect(failures.Exitcode).To(Equal(6))
+		}, "30s", "1s").Should(Succeed())
```
Comment (jakubdyszkiewicz) [inline test/e2e/reachableservices/auto_reachable_mesh_services_k8s.go:119]: 30s? on non-env e2e suite? That's a lot. Can we make it 5s?

File: pkg/core/resources/apis/mesh/dataplane_validator.go [hunk 15]
```diff
@@ -257,6 +263,33 @@ func validateOutbound(outbound *mesh_proto.Dataplane_Networking_Outbound) valida
...
+			default:
+				result.AddViolationAt(path.Index(i).Field("kind"), fmt.Sprintf("invalid value. Available values are: %s", strings.Join(maps.SortedKeys(allowedKinds), ",")))
+			}
+			if backendRef.Name == "" && backendRef.Namespace == "" && len(backendRef.Labels) == 0 {
+				result.AddViolationAt(path.Index(i).Field("name"), "name or labels are required")
+			}
+			if backendRef.Name == "" && backendRef.Namespace != "" {
+				result.AddViolationAt(path.Index(i).Field("name"), "name is required, when namespace is defined")
+			}
+			if s := strings.Split(backendRef.Name, "."); len(s) > 1 {
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/core/resources/apis/mesh/dataplane_validator.go:282]: We can allow creating the name with a `.` but at this level (validator) we cannot search for the mesh service. I think I can remove the validation

File: pkg/xds/context/context.go [hunk 16]
```diff
@@ -59,27 +60,39 @@ func (g BaseMeshContext) Hash() string {
 	return base64.StdEncoding.EncodeToString(g.hash)
 }
 
+type LabelValue struct {
+	Label string
+	Value string
+}
+type (
+	ServiceName                  string
+	LabelsToValuesToServiceNames map[LabelValue]map[ServiceName]bool
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply pkg/xds/context/context.go:69]: We agreed that it makes sense to add it in a separate PR

File: pkg/plugins/policies/meshtrafficpermission/graph/reachable_graph.go [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+	rule := r.rules[toTags[mesh_proto.ServiceTag]].Compute(core_rules.SubsetFromTags(fromTags))
+	if rule == nil {
+		return false
+	}
+	action := rule.Conf.(mtp_api.Conf).Action
+	return action == mtp_api.Allow || action == mtp_api.AllowWithShadowDeny
+}
+
+func (r *Graph) CanReachBackend(fromTags map[string]string, backendRef *mesh_proto.Dataplane_Networking_Outbound_BackendRef) bool {
+	if backendRef.Kind == "MeshExternalService" {
```
Comment (lobkovilya) [inline pkg/plugins/policies/meshtrafficpermission/graph/reachable_graph.go:40]: could you please use the constant here?

File: pkg/xds/sync/dataplane_proxy_builder.go [hunk 18]
```diff
@@ -136,6 +141,32 @@ func (p *DataplaneProxyBuilder) resolveVIPOutbounds(meshContext xds_context.Mesh
...
+				}
+				// check if there is an entry with specific port or without port
+				if !pointer.Deref(reachableBackends)[backendKey] && !pointer.Deref(reachableBackends)[xds_context.BackendKey{Kind: outbound.BackendRef.Kind, Name: outbound.BackendRef.Name}] {
+					// ignore VIP outbound if reachableServices is defined and not specified
+					// Reachable services takes precedence over reachable services graph.
+					continue
+				}
+				// we don't support MeshTrafficPermission for MeshExternalService at the moment
+				// TODO: https://github.com/kumahq/kuma/issues/11077
+			} else if outbound.BackendRef.Kind != "MeshExternalService" {
```
Comment (lobkovilya) [inline pkg/xds/sync/dataplane_proxy_builder.go:164]: and here

File: pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+		rl, ok, err := graph_util.ComputeMtpRulesForTags(dpTags, trimNotSupportedTags(mtps, dpTags))
+		if err != nil {
+			log.Error(err, "service could not be matched. It won't be reached by any other service", "service", ms.Meta.GetName())
+			continue
+		}
+		if !ok {
+			continue
+		}
+		rules[BackendKey{
+			Kind: "MeshService",
```
Comment (lobkovilya) [inline pkg/plugins/policies/meshtrafficpermission/graph/backends/reachable_backend_refs_graph.go:41]: and here

**Review by lobkovilya — APPROVED** [review]

**Pull request merged**